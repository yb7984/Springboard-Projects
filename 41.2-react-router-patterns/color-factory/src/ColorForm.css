.ColorForm {
    background-color: blanchedalmond;
    width: 100vw;
    height: 100vh;
    font-size: 1.5em;
    padding-top: 100px;
}

.ColorForm-field{
    text-align: left;
    max-width: 400px;
    margin: 0px auto;
    padding:20px;
}

.ColorForm-field input{
    font-size: 1.5em;
    width: 100%;
}
.ColorForm button{
    font-size: 1.2em;
    border-radius: 10px;
    padding:10px 30px 10px 30px;
}