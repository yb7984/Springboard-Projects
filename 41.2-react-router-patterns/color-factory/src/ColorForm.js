import { useState } from "react";
import {Link} from 'react-router-dom';
import './ColorForm.css';

const ColorForm = ({ colors, addColor }) => {

    const INITIAL_STATE = {
        name: "",
        value: "#FFFFFF"
    };
    const [formData, setFormData] = useState(INITIAL_STATE);

    const handleSubmit = (e) => {
        e.preventDefault();

        if (colors.find(color => (color.name === formData.name))) {
            alert("Color Name has been used, please choose another one!");
            return;
        }

        addColor(formData.name, formData.value);

        setFormData(INITIAL_STATE);
    };

    const handleOnChange = (e) => {
        const { name, value } = e.target;

        setFormData(fData => ({
            ...fData,
            [name]: value
        }));
    };

    return (
        <div className="ColorForm">
            <form onSubmit={handleSubmit}>
                <div className="ColorForm-field">
                    <label htmlFor="name">Color Name:</label>
                    <input type="text" id="name" name="name" value={formData.name} onChange={handleOnChange} required />
                </div>
                <div className="ColorForm-field">
                    <label htmlFor="value">Color Value:</label>
                    <input type="color" id="value" name="value" value={formData.value} onChange={handleOnChange} required />
                </div>
                <button>Add Color!</button>
            </form>
            <h1>
                <Link to="/colors">Go back!</Link>
            </h1>
        </div>
    )
}

export default ColorForm;