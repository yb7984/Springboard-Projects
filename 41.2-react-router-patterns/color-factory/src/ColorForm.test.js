import ColorForm from './ColorForm';
import { fireEvent, render } from '@testing-library/react';
import { MemoryRouter } from 'react-router';

let colors = [
    {
        name: "Red",
        value: "#FF0000"
    }
];

//mock the addColor function
const addColor = (name, value) => {
    colors.push({
        name,
        value
    })
};

const resetColors = () => {
    colors = [
        {
            name: "Red",
            value: "#FF0000"
        }
    ];
}

let alertMsg = null;

window.alert = (msg) => {
    alertMsg = msg;
}


it('renders without crashing', () => {
    render(<MemoryRouter><ColorForm colors={colors} /></MemoryRouter>);
});


it("matches snapshot", function () {
    const { asFragment, queryByLabelText } = render(
        <MemoryRouter>
            <ColorForm colors={colors} />
        </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();

    expect(queryByLabelText('Color Name:')).toBeInTheDocument();
    expect(queryByLabelText('Color Value:')).toBeInTheDocument();
});


it("works with adding new color not in the list", function () {
    resetColors();

    const { queryByLabelText, queryByText } = render(
        <MemoryRouter>
            <ColorForm colors={colors} addColor={addColor} />
        </MemoryRouter>
    );

    const inputName = queryByLabelText('Color Name:');
    const inputValue = queryByLabelText('Color Value:');
    const btn = queryByText("Add Color!");
    fireEvent.change(inputName, { target: { value: "Blue" } });
    fireEvent.change(inputValue, { target: { value: "#0000ff" } });

    fireEvent.click(btn);

    expect(colors[1]).toEqual({ name: "Blue", value: "#0000ff" });
});


it("doesn't work with adding new color already in the list", function () {
    resetColors();

    const { queryByLabelText, queryByText } = render(
        <MemoryRouter>
            <ColorForm colors={colors} addColor={addColor} />
        </MemoryRouter>
    );

    const inputName = queryByLabelText('Color Name:');
    const inputValue = queryByLabelText('Color Value:');
    const btn = queryByText("Add Color!");
    fireEvent.change(inputName, { target: { value: "Red" } });
    fireEvent.change(inputValue, { target: { value: "#0000ff" } });

    fireEvent.click(btn);

    expect(colors[1]).toEqual(undefined);
    expect(alertMsg).toEqual("Color Name has been used, please choose another one!");

    alertMsg = "";
});

