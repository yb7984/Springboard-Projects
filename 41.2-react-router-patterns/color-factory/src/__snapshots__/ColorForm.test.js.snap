// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ColorForm"
  >
    <form>
      <div
        class="ColorForm-field"
      >
        <label
          for="name"
        >
          Color Name:
        </label>
        <input
          id="name"
          name="name"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="ColorForm-field"
      >
        <label
          for="value"
        >
          Color Value:
        </label>
        <input
          id="value"
          name="value"
          required=""
          type="color"
          value="#FFFFFF"
        />
      </div>
      <button>
        Add Color!
      </button>
    </form>
    <h1>
      <a
        href="/colors"
      >
        Go back!
      </a>
    </h1>
  </div>
</DocumentFragment>
`;

exports[`matches snapshot 2`] = `
<DocumentFragment>
  <div
    class="ColorForm"
  >
    <form>
      <div
        class="ColorForm-field"
      >
        <label
          for="name"
        >
          Color Name:
        </label>
        <input
          id="name"
          name="name"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="ColorForm-field"
      >
        <label
          for="value"
        >
          Color Value:
        </label>
        <input
          id="value"
          name="value"
          required=""
          type="color"
          value="#FFFFFF"
        />
      </div>
      <button>
        Add Color!
      </button>
    </form>
    <h1>
      <a
        href="/colors"
      >
        Go back!
      </a>
    </h1>
  </div>
</DocumentFragment>
`;

exports[`works with adding new color not in the list 1`] = `
<DocumentFragment>
  <div
    class="ColorForm"
  >
    <form>
      <div
        class="ColorForm-field"
      >
        <label
          for="name"
        >
          Color Name:
        </label>
        <input
          id="name"
          name="name"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="ColorForm-field"
      >
        <label
          for="value"
        >
          Color Value:
        </label>
        <input
          id="value"
          name="value"
          required=""
          type="color"
          value="#FFFFFF"
        />
      </div>
      <button>
        Add Color!
      </button>
    </form>
    <h1>
      <a
        href="/colors"
      >
        Go back!
      </a>
    </h1>
  </div>
</DocumentFragment>
`;
